### ISSUE TITLE
[HIGH] Add SQLite support to web browser and fallback to IndexedDB

### ISSUE THREAD

**Issue by muttmuure** [Expensify/App#34289]
Placeholder tracking issue for adding SQLite support to Chrome, Safari, Edge and FireFox (Not IE) 

Also for adding a fallback mechanism to retrieve data from IndexedDB when SQLite fails or has a problem

cc @roryabraham

**Comment by muttmuure** [comment]
Hey @MIAhmed! What's your Slack handle? Would you be interested in creating a proposal for this?

**Comment by roryabraham** [comment]
Thanks for running with this @muttmuure, I think generally there was consensus that this is a good idea, but let me post a formal proposal first just to get everyone on board. Going to slap a planning label on this and put it on HOLD

**Comment by roryabraham** [comment]
Chatting in [this thread](https://expensify.slack.com/archives/C05LX9D6E07/p1704921473038539)

**Comment by adhorodyski** [comment]
Does this still rely on the [deprecated WebSQL](https://www.w3.org/TR/webdatabase/) standard? As far as I'm aware only WASM is now a viable option to use SQL on this platform.

**Comment by MIAhmed** [comment]
> Hey @MIAhmed! What's your Slack handle? Would you be interested in creating a proposal for this?
@muttmuure Yes, I am interested in implementing fallback to the indexedDB as I have already implemented that.
I do not have Slack access

**Comment by roryabraham** [comment]
> Does this still rely on the [deprecated WebSQL](https://www.w3.org/TR/webdatabase/) standard? As far as I'm aware only WASM is now a viable option to use SQL on this platform.

No, I'm referring to the official SQLite wasm build

**Comment by roryabraham** [comment]
Sounds like we got consensus to do some exploratory research here and investigate the potential performance benefits.

**Comment by roryabraham** [comment]
For transparency to our open-source community, we've asked Margelo to begin researching this since they're experienced with working in Onyx and maintain https://github.com/Margelo/react-native-quick-sqlite

**Comment by chrispader** [comment]
commenting here for assignment! @roryabraham

**Comment by chrispader** [comment]
I did some research today and here's are my initial thoughts: cc @roryabraham @muttmuure @adhorodyski 

(Lmk if you have any comments/critic)

## Which SQLite/WASM approach

As suggested in the documentation of the official library ([SQLite WASM](https://github.com/sqlite/sqlite-wasm)) i would go for the approach where we use the extra wrapped worker thread. This way, additionally to (potentially) improved speed we have the benefit of offloading much of the load in Onyx to a different thread.

Reference: [Worker + Promise approach](https://sqlite.org/wasm/doc/tip/api-worker1.md)

> @chrispader having set up wasm support for Encryptify, I wonder if this is something we can implement at the react-native-quick-sqlite layer?

I would definitely add this functionality to `react-native-quick-sqlite` directly. The library is not yet web-compatible anyway, so this would simply extend the library by adding the whole web part and wouldn't interfere with the native implementation at all.

## Onyx integration

> That way, we would be able to leave Onyx _mostly_ unchanged – just check if OPFS is supported, and if it is use the existing quick-sqlite storage provider, if not fall back on IndexedDB.

Exactly! As you've suggested, we would then check for availability of OPFS in Onyx and use `IndexedDB` (`idb-keyval` provider) as fallback.

Onyx operates async anyways, so this approach wouldn't require any changes in the library.

From https://expensify.slack.com/archives/C05LX9D6E07/p1704972138886149?thread_ts=1704921473.038539&cid=C05LX9D6E07 by @adhorodyski 
> one of the benefits is the unification of the query layer - this can potentially really simplify the architecture which is now limited to the least capable platform (key-value on web). Think indexes for example, running migrations, or bringing in an ORM for a type-safe database access.

I'm not sure if i understood your idea 100%, but if it refers to re-designing Onyx to support relational data, i don't think that would be a good idea atm. I wouldn't suggest to rely on any specific storage provider implementation too much, like SQLite on both web and native. Or was the idea to drop Onyx and create a completely re-designed (relational data) storage solution? 

Imo Onyx is exclusively designed as a key-value store and its big advantage is that it's not dependent on any specific storage layer, instead it can use any underlying storage solution underneath... and this will potentially even improve even more.

I also agree with @tgolen, that as long as we don't completely re-structure our approach to querying and persisting data, i don't think we need `react-query` right now.

(Correct me if i got any aspect of your suggestion wrong though)

## Performance

As @roryabraham already mentioned in [this Slack thread](https://expensify.slack.com/archives/C05LX9D6E07/p1704921473038539) the main benefit would definitely be faster speeds and to offload a lot of work to a separate worker thread and therefore unblock the main thread and improve app's performance.

I haven't created any real benchmarks and performance comparisons with IndexedDB yet, but from what i've read it's definitely faster than `IndexedDB`.

## Analytics

> We could also create a log of grafana graph to see how much the IndexedDB provider is actually used in the wild. That way a year from now there's no guesswork – we can look at how much it's actually used by users and consider dropping it from Onyx to simplify the codebase.

That definitely makes sense! Do we already use Grafana and is there an account/token to use it?

I could also imagine adding analytics for web and native performance of SQLite. This way we could analyze which kind of  devices profit the most from this change and which might have problems.

## Going forward...

@roryabraham @tgolen just to make sure i'm not putting too much work into this straight ahead, was the idea to create a P/S for this with all the actual problems solved in the app first?

Or do we want to dive straight into the actual implementation in `react-native-quick-sqlite`? After my initial research, i don't see any functionality of the library that couldn't be realized with the WASM implementation of SQLite.

**Comment by chrispader** [comment]
As for the implementation:

`react-native-quick-sqlite` is currently only synchronous. Also, the WASM API is limited, though totally sufficient for the use in Onyx/Expensify.

My idea would be to create a separate version inside the `react-native-quick-sqlite` library, which is

- asynchronous and
- limited in functionality, based on what the WASM implementation allows

This way, our other users of the library aren't affected and if the WASM implementation of SQLite get's more comprehensive in the future, we can extend the scope of the new version and also provide an additional async version.

**Comment by roryabraham** [comment]
Thanks @chrispader, here are my thoughts:

> Was the idea to create a P/S for this with all the actual problems solved in the app first? ... Or do we want to dive straight into the actual implementation in react-native-quick-sqlite?

Basically, I think we should try to do the minimum amount of work possible to get real benchmarks that show the difference in performance between IndexedDB and SQLite-wasm. It would also be interesting to see how offloading Onyx work to a separate worker thread would help speed up the app, but I don't see how we could investigate that without building out a complete solution.

Once we have benchmarks that show the performance change, we can evaluate whether it's worth pursuing further. I admit it's difficult to write a problem statement for performance improvements because "how fast is _fast enough_"? I think a major selling point for us though is that the amount of work in Onyx and E/App to make this change is _very small_ – all the work prettymuch is in react-native-quick-sqlite.

> react-native-quick-sqlite is currently only synchronous

Maybe for simplicity start with the synchronous SQLite-wasm implementation then? I'd think that the way to approach this would be:

- Start with the easiest thing, which is to create a synchronous implementation for web to match the synchronous implementation on iOS and Android.
- Then build an async implementation for web, iOS, and Android
- Then, if it's beneficial, update Onyx to use the async implementation.

That way there's no breaking changes, and all platforms have both sync and async capabilities, without any confusing configuration, sub-packages, or patches

**Comment by quinthar** [comment]
Just confirming that we have @danielk1977 and @drhsqlite in this discussion, as they are the literal experts!

**Comment by danielk1977** [comment]
I'm here. But @sgbeal is the sqlite/wasm expert.

**Comment by sgbeal** [comment]
> As suggested in the documentation of the official library ([SQLite WASM](https://github.com/sqlite/sqlite-wasm)) i would go for the approach where we use the extra wrapped worker thread. This way, additionally to (potentially) improved speed we have the benefit of offloading much of the load in Onyx to a different thread.
> 
> Reference: [Worker + Promise approach](https://sqlite.org/wasm/doc/tip/api-worker1.md)

If i may interject here, as the person who wrote that documentation and code: The worker1 API is _not_ recommended for any serious work. It has some severe limitations, noted in the above link, e.g. making it work properly with transactions is difficult and it cannot handle nested loops of queries because JS's postMessage() model simply can't support it.

The sqlite/wasm library can, and arguably should, be run in a separate thread _without_ the worker1 API. The most flexible way to use it in a separate thread is to load the library and any directly-db-using client code into the same worker. That gives the client code full access to the sqlite API without the limitations of the worker1 API. It's not only much more comfortable to use that way, but also much more powerful. The one down-side is that the client-side code on the other side of the worker will have to coordinate its own communication do the db-using worker. That also has advantages, however, such as being able to encapsulate all of the SQL in the db-using thread and accessing it via a higher-level postMessage() API from the main thread.

**Comment by sgbeal** [comment]
> I haven't created any real benchmarks and performance comparisons with IndexedDB yet, but from what i've read it's definitely faster than `IndexedDB`.

We have never worked with IndexedDB and have no benchmarks comparing it. We do have a wasm port of sqlite3's native-level benchmarking tool, speedtest1, which can be used to get a fairly good idea of its overall speed:

The baseline is an in-memory db:

<https://wasm-testing.sqlite.org/speedtest1-worker.html?size=15>

Then we have two alternatives for persistence via OPFS, each with their own trade-offs, as documented [here](https://sqlite.org/wasm/doc/trunk/persistence.md):

- <https://wasm-testing.sqlite.org/speedtest1-worker.html?vfs=opfs&size=15>
- <https://wasm-testing.sqlite.org/speedtest1-worker.html?vfs=opfs-sahpool&size=15>

(That "size" URL parameter can be used to adjust the workload size - note the query counts in the output.)

I/O of the former is significantly slower than the latter but it offers a moderate level of concurrency (multi-tab use), whereas the latter is blazing fast but cannot be accessed from more than one tab/window at a time. Note that "slow" and "fast" are always relative, however, and the former is "fast enough" for many types of applications.

Note that concurrency in OPFS is a pain point - it does not natively support _any_ concurrency without the use of experimental new APIs which were only recently released and are not available outside of Chrome. Our first VFS works around those limitations somewhat but it cannot reliably support more than a few browser tabs visiting the same db without locking-related issues. Our second VFS sacrifices all concurrency support for speed, and a given pool of dbs can only be accessed by a single tab concurrently.

For the sake of transparency, i feel compelled to point out that our wasm build  is _not_ the only option. [Roy Hashimoto's wa-sqlite project](https://github.com/rhashimoto/wa-sqlite) is actively developed, supports multiple storage back-ends (including IndexedDB) via the sqlite VFS API, and is specifically developed for the cutting-edge JS ecosystem. In contrast, we publish plain-vanilla JS and make no use of node.js-based toolchains, which has been a point of contention with some potential users but it's a stance we must stick to for the sake of developer bandwidth and sanity.

**Comment by chrispader** [comment]
> If i may interject here, as the person who wrote that documentation and code: The worker1 API is _not_ recommended for any serious work. It has some severe limitations, noted in the above link, e.g. making it work properly with transactions is difficult and it cannot handle nested loops of queries because JS's postMessage() model simply can't support it.

Ah ok, thanks for the notice! By the "suggested approach in the official documentation" i was mostly referring to the [GH Readme](https://github.com/sqlite/sqlite-wasm#in-a-wrapped-worker-with-opfs-if-available) where it says "preferred option 🏆). If you suggest to use a differnet, i'd be glad to go with that instead!

> The sqlite/wasm library can, and arguably should, be run in a separate thread _without_ the worker1 API. The most flexible way to use it in a separate thread is to load the library and any directly-db-using client code into the same worker. 
> ...
> The one down-side is that the client-side code on the other side of the worker will have to coordinate its own communication do the db-using worker. That also has advantages, however, such as being able to encapsulate all of the SQL in the db-using thread and accessing it via a higher-level postMessage() API from the main thread.

If i understand you correctly here, this would mean that we can't or shouldn't implement this directly as-planned into `react-native-quick-sqlite` and indirectly into `react-native-onyx`, because we also need client-code from `Expensify/App` to run in the worker thread and directly communicate with the database.

I mean there is probably some hacky way to achieve this with all sorts of callbacks, but that already sounds like callback hell...

(@sgbeal please correct me if anything in this statement is incorrect or if i am mistaken somehow)

cc @roryabraham

**Comment by chrispader** [comment]
Also JFYI @roryabraham

I'm going to be 100% OOO from 31/01/2024 - 17/03/2024. From what i can see right now, the work on this feature will either take longer than anticipated (from my side) and/or involve much more than just a change to `react-native-quick-sqlite`.

Therefore, i can finish my research on this and we can figure out the best approach to achieve this feature and then i'd have to hand this task over to another Margelo engineer. Maybe @Szymon20000 or @janicduplessis, since they also both have experience with `react-native-onyx` and/or `react-native-quick-sqlite`.

I'm happy to put more work into research, benchmarking and design of implementation until then, but i think the comment from @sgbeal changes a lot about my initial proposal and our initial idea of implementing this.

**Comment by sgbeal** [comment]
> If i understand you correctly here, this would mean that we can't or shouldn't implement this directly as-planned into `react-native-quick-sqlite` and indirectly into `react-native-onyx`, because we also need client-code from `Expensify/App` to run in the worker thread and directly communicate with the database.

My experience with React is literally zero, so i can't speculate on what is or is not feasible with it, but loading both sqlite and your db-side client code into a single thread should not limit what you can do, beyond any limitations which you build in to that abstraction.

Let's say you've got a primary thread (be it the main window thread or a worker - that's largely irrelevant for sqlite), then load the library _and_ your db-side code into its own shared thread. That gives your db-side code full access to the sqlite API (both the C-style API and a higher-level OO APIs), with no limitations on what it can do vis a vis the sqlite API. Your primary thread could then postMessage() requests to your client-level code in the db thread, and you can react to that however you like, within the limitations of the postMessage() protocol (e.g., iterating over nested loops of queries is not possible with that approach, and transactions are tricky to get right).

Your primary thread can post a message like:

```
postMessage({type: 'getDataForCustomer', customerId: 123456 })
```

and your db thread will have full access to the sqlite API, including custom SQL functions, to fetch that data, reformat it into whatever form you want the primary thread to have, and return it via a response postMessage(). The supplied worker1 API and its "promiser" proxy simplify that interaction somewhat, but its model of working introduces several significant limitations, such as:

- Inability to use nested query loops. That is, nested for(...) loops which each run a query are not possible in that model. Subselects within a single query are unaffected.
- Inability (or nearly so) to use transactions unless they can be packed into a single SQL string (containing an arbitrary number of statements), noting that the large-blob-of-SQL approach precludes the use of bind() for binding SQL parameters.
- Inability to install custom SQL functions.

Plus the SQL will all have to be part of your primary thread, rather than separated neatly into the db-using thread.

Even so, it would not require a significant development effort to write your own custom worker1-like proxy which is specific to your APIs and moves all of the db-related logic into the db worker, which your primary thread then remote-controls via a custom-fit API. That would enable you to eliminate the above limitations because they would not apply to that hypothetical new remote-controlled worker.

> I mean there is probably some hacky way to achieve this with all sorts of callbacks, but that already sounds like callback hell...

Working via postMessage() always has a bit of callback hell involved, even through the callbacks are effectively encapsulated in the onmessage handler. The worker1 promiser proxy effectively uplifts those callbacks from onmessage into something more recognizable as callback hell, but it's still cross-thread callback hell, and the cross-thread aspect of it introduces the above-mentioned significant usage limitations. Those APIs were created as proxies for _simple_ use cases, and suffice for certain types of applications, but your apps would almost certainly outgrow those capabilities within days.

You won't be able to avoid postMessage() altogether unless you have sqlite in your main app thread, but OPFS is not available when sqlite is run that way. We can, however, simplify your use of postMessage() by housing your db-using code and sqlite in the same worker thread and remote-controlling that combination from your primary thread.

> I'm happy to put more work into research, benchmarking and design of implementation until then, but i think the comment from @sgbeal changes a lot about my initial proposal and our initial idea of implementing this.

Likewise, we're happy to work with you to recommend a best-fit approach, with the caveat that our collective React know-how and ambitions are literally zero. Since the JS ecosystems evolves so rapidly, beyond our tiny team's ability to keep up with or sensibly claim to support, we very specifically provide only plain-vanilla JS which we then expect interested parties to integrate into their favorite framework as they see fit. We can't assist much with the framework-specific details but are happy to assist with both the integration and how to make use of the JS APIs.

We're available just about any time for video calls, if you'd like to talk through this.

**Comment by adhorodyski** [comment]
> Relational data approach / Onyx re-design?
> 
> 
> From https://expensify.slack.com/archives/C05LX9D6E07/p1704972138886149?thread_ts=1704921473.038539&cid=C05LX9D6E07 by @adhorodyski
> 
> > one of the benefits is the unification of the query layer - this can potentially really simplify the architecture which is now limited to the least capable platform (key-value on web). Think indexes for example, running migrations, or bringing in an ORM for a type-safe database access.
> > 
> 
> I'm not sure if i understood your idea 100%, but if it refers to re-designing Onyx to support relational data, i don't think that would be a good idea atm. I wouldn't suggest to rely on any specific storage provider implementation too much, like SQLite on both web and native. Or was the idea to drop Onyx and create a completely re-designed (relational data) storage solution?
> 
> I like the idea of improving the overall performance by not trying to convert relational data into key-value data and therefore losing potential of indexing and efficient querying. I just think it wouldn't be a good idea to try achieve this in Onyx...
> 
> Imo Onyx is exclusively designed as a key-value store and its big advantage is that it's not dependent on any specific storage layer, instead it can use any underlying storage solution... and we will potentially improve this aspect even more.
> 
> I also agree with @tgolen, that as long as we don't completely re-structure our approach to querying and persisting data, i don't think we need `react-query` right now.
> 
> (Correct me if i got any aspect of your suggestion wrong though)
> 

also @cc @dsilva as you might be interested in this:

Hey @chrispader, you totally got this right and I wanted to shed some light onto the React side of things and how SQLite in the long run can be integrated with the UI layer, not Onyx-related. My idea is definitely not restructuring Onyx and I agree, react-query doesn't fit nowhere in there currently :)

What I rather wanted to propose is **establishing a long-term vision for a new solution that is really scalable and future proof for the app's performance at any scale. This includes relying on the relational data structures and the power of SQL as well as the latest concurrency features available in React.**

### The problem

Currently, we rely on the key-value (KV) architecture even for mobile devices (running SQLite) which forces us to calculate all the bits of information on the same thread that is responsible for handling user interactions. No matter where we got the data from (be it indexedb or an sqlite table), as the scale increases, we have to add up another iterations and use more of the device resources to handle the workload.
For reference, please see the below screenshot showing how this looks like for an Android device on the JS thread, with Heavy account when sending a message (one of the core metrics):

<img width="1102" alt="defaultProductionHermesTrace" src="https://github.com/Expensify/App/assets/42047036/cd8d6063-8942-4595-bd89-76c7590d083a">

As we can see, most of the time (here ~15 seconds) is being spent on resorting list items, **completely blocking user interactions**, and it’s like this by design - with KV we can do all the tricks, but every filter operation, fields recalculation and all the heavy lifting still sits on the JS engine.

This can be optimised, but only to a certain point at which we won't be able to squeeze anything more out of it without sacrificing a few seconds for the most basic operations. I’m sure this is nowhere we want to be a few years from now.

What's the most important, **this pattern spreads across our core metrics (App startup, Opening up a report, Sending a message)**, making it really hard (if not impossible at some point) to hit the performance targets as we scale.

### Solution

Even though we have SQLite, we don’t currently use nearly any of its features. We utilise it as a KV.

**For the long term, as SQLite becomes available also for the web platform, I propose utilising the tool at hand to its full potential, enabling us to offload all the data related calculations in the App to the database through SQL.**

Efficient querying with filters, joining tables, sorting (ordering), indexing or running migrations can all run outside of the JS thread within milliseconds.

An example from before running on the database through `ORDER BY` should take only some milliseconds and be instantly available to the UI.

Moreover, all of this can operate with some really minimal tooling around as React aims to fully support Promises with the concurrent features - and even the most complex query running through a database driver *is just a Promise*. This means that we can replace Onyx with a really thin and efficient communication layer between the UI and DB and also provide a good developer experience - will less code to write, easy to follow patterns and minimalistic abstractions.

There are also a few projects which currently try to address these problems in a similar way:

https://www.evolu.dev/

I really wanted to open this up for some feedback, as I’ve heard there’s an interest in utilising SQLite more as we get the web support in place.

**An offline-first application should be no slower than tens of milliseconds per operation basically no matter the scale - and that’s what we can definitely aim for.**

**Comment by quinthar** [comment]
We are huge huge fans of SQLite.  Can you outline a strategy for how to incrementally migrate Onyx to use more SQLite functionality, without needing to rewrite everything or lose momentum?  You've mentioned that grouping, sorting, and filtering are all great features of a relational database, and that doing these after the fact in the application layer is super slow.  Maybe you could pick a couple scenarios that are particularly slow and suggest how in theory we could move those slow operations into the SQLite layer, without fundamentally breaking the Onyx API layer?

**Comment by quinthar** [comment]
Escalating this conversation to Slack to make it go faster: https://expensify.slack.com/archives/C01GTK53T8Q/p1706933222963169

**Comment by roryabraham** [comment]
Left a [proposal in the slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1707265979464169?thread_ts=1706933222.963169&cid=C01GTK53T8Q) to try and gain consensus on next steps here.

**Comment by roryabraham** [comment]
Ah, @drhsqlite that's a link to our #expensify-open-source slack room. We can invite you if you'd like, but the proposal is not specific with regards to the implementation, just high-level next steps

<img width="510" alt="image" src="https://github.com/Expensify/App/assets/47436092/4deef821-49e2-4475-b8d3-ab67c582592e">

Most of the links are internal, but you can see some previous benchmark results comparing two competing IndexedDB storage solutions [here](https://github.com/Expensify/react-native-onyx/pull/293#issue-1835213896)

**Comment by hannojg** [comment]
Yes, happy to add it to the existing performance benchmark tool and get us some numbers!

**Comment by kirillzyusko** [comment]
Hey, I am helping with implementing the performance benchmark, please assign me as well. Thanks!

**Comment by kirillzyusko** [comment]
Hello everyone 👋 

These are preliminary benchmark results:

|Operation       |Baseline (mock storage)|IDB (current)|Local forge|SQLite (main)|SQLite (worker)|
|----------------|-----------------------|-------------|-----------|-------------|---------------|
|Single set      |115.9                  |157.8        |170.1      |155.4        |287.6          |
|Get collection  |10.39                  |9.46         |10.66      |9.55         |10.39          |
|Merge collection|365.7                  |619.8        |61338.8    |542.7        |600.8          |
|Single merge    |114.7                  |185.6        |192        |192.4        |311.2          |
|Clear           |99.08                  |982.7        |681        |166.4        |612            |

The agenda is:
- **_Baseline (mock storage)_** - is not a storage at all. All functions are immediately returns `Promise.resolve`. Added it to see how much time is needed for Onyx to actually call necessary functions.
- **_IDB (current)_** - current storage that used in Expensify
- **_Local forge_** - previous storage that was replaced by IDB
- **_SQLite (main)_** - SQLiteWASM running on a main thread (🚨 note, that in this case it will be **in-memory** DB, and all data will be lost after page reload - it happens because OPFS is not available on a main thread 🚨).
- **_SQLite (worker)_** - SQLiteWASM running on a worker thread (with OPFS)

All metrics were captured with `depth=1, length=35, items=10_000` and all operations were performed `10x` times and then average value was calculated `(sum / 10)`.

The working branch is: https://github.com/margelo/react-native-onyx/tree/feat/sqlite-wasm

As I can see **single merge** and **single set** operations are slower comparing to IDB - I think it's mainly due to the fact that we are passing large JSON objects into `postMessage` and data serialization steals some CPU time. I'll investigate it further whether it's possible to optimize it or not 👀

**Comment by tgolen** [comment]
Thanks. Are those values in milliseconds?

From those preliminary results, it doesn't look like there are any significant gains, and maybe even some losses. JSON serialization is the worst :rage4: 

Hopefully, there is a way to optimize it.

**Comment by kirillzyusko** [comment]
> Are those values in milliseconds?

Yes, in milliseconds.

> Hopefully, there is a way to optimize it.

I'll try to find possible ways 👀

**Comment by sgbeal** [comment]
> SQLite (worker) - SQLiteWASM running on a worker thread (with OPFS)

Note that there are two distinct OPFS VFSes. One of them trades some level of concurrency (multi-tab access) for a significant cost in speed and the second forgoes all concurrency for corresponding speed gains (it's typically 2-3x faster than the other one, depending on the workload, and is never slower). Those VFSes can be run through a wasm build of sqlite's "speedtest1" benchmark, the project's own I/O-centric benchmarking tool, via:

https://wasm-testing.sqlite.org/

See the "speedtest1" links around the center of the page for the various options, including in-memory, "opfs" (the first OPFS VFS), "opfs-sahpool" (the second OPFS VFS), and "kvvfs" (which can use sessionStorage and localStorage for _small_ amounts of persistence in the main thread).

Chrome v121 has a new, experimental, still-up-for-change feature which _should_ permit us to add concurrency support to the second (faster) VFS, but (A) that's still a big question mark and (B) it's currently only in bleeding-edge Chrome versions (and subject to change).

The two current OPFS VFSes are covered at:

https://sqlite.org/wasm/doc/trunk/persistence.md#opfs

**Comment by kirillzyusko** [comment]
Thanks for pointing out @sgbeal 

I used `new sqlite3.oo1.OpfsDb(`/${DB_NAME}.sqlite3`);` construction, I think by default it resolves to `vfs=opfs` (i. e. lower speed, concurrency support), right?

Anyway I updated table (I added benchmark only for SQL queries execution) and in this case SQLite shows the fastest results. So I think the problem here is in data serialization between threads. I'm investigating whether it can be improved somehow 👀

**Comment by danieldoglas** [comment]
Quick noob question: 

based on the speed test in https://wasm-testing.sqlite.org/speedtest1.html, if instead of storing JSONs we had a few structured tables for data that will be abundant and we have a format that is mostly defined (reportActions, transactions) would those operations be faster?

**Comment by kirillzyusko** [comment]
> if instead of storing JSONs we had a few structured tables for data that will be abundant and we have a format that is mostly defined (reportActions, transactions) would those operations be faster?

@danieldoglas SQLite is also fast with plain JSON messages. The main bottleneck right now is a communication between threads (main <--> worker).

**Comment by kirillzyusko** [comment]
> Regarding the byte-level serialization: rather than postMessage() data back and forth between the VFS-specific thread and the library thread, the VFS thread communicates with the library's thread over a SharedArrayBuffer, so there is no extraneous copying going on between those threads. The two threads use Atomics.notify() to make each other aware of new data in the buffer. The communication between the library and the VFS always happens in chunks equal to the db page size, so the VFS can (and does) reserve a fixed-sized SharedArrayBuffer for its communication with the library. The major performance hit in all of that is us waiting on the cross-thread communication between Atomics.notify() and Atomics.wait(), all of which happens in black-box browser internals.

@sgbeal so you are saying, that usage `SharedArrayBuffer` should make communication faster? I thought about experimenting with it, but still kind of afraid, that writing/reading from `SharedArrayBuffer` can be a bottleneck 🤔 But I'll try to see how it goes! 👀

**Comment by sgbeal** [comment]
> based on the speed test in https://wasm-testing.sqlite.org/speedtest1.html, if instead of storing JSONs we had a few structured tables for data that will be abundant and we have a format that is mostly defined (reportActions, transactions) would those operations be faster?

That's really anyone's guess and mine would be mere speculation. That will, at least in part, depend on whether those objects span db pages, because...

> @sgbeal so you are saying, that usage SharedArrayBuffer should make communication faster?

Faster compared to (say) postMessage() because it eliminates an extra copy which might otherwise be necessary[^1]. For example, when the library reads from storage, it passes a destination byte array to the VFS. In this particular case, the library and the VFS-related I/O live in different threads but they communicate via the same SharedArrayBuffer, so the I/O itself is performed directly on that shared buffer. The coordination/synchronization of the communication between those two threads, irrespective of the actual I/O, is where the biggest single performance hit comes from. The process goes something like:

- The library tells the VFS "go fetch this byte range while I wait on you," at which point the library uses Atomics.wait() to wait on completion.
- The I/O thread, which has been using Atomics.wait() to wait for a task request, wakes up, reads those bytes into the SharedArrayBuffer, reports "work is done - you may continue now" using Atomics.notify(). It then calls Atomics.wait() to wait on the next request.
- The library thread receives that Atomics.notify() notification via its own blocking Atomics.wait() call and can then consume the bytes from the SharedArrayBuffer.

The same happens for writing: the library thread populates the SAB, tells the I/O thread that data is available, and then waits on the I/O thread's response to that write. Thanks to the SAB, the I/O thread does not need to make a separate copy of that data before passing it on to the OPFS layer.

Does that answer your question?

[^1]: postMessage() _can_ move, rather than copy, buffers between threads, but in our case doing so is more expensive/slow than an Atomics.wait/notify()-based construct because we have to either create a new buffer for each operation or coordinate moving of a single buffer back and forth between threads. Our very first prototype used postMessage() for the cross-thread communication but it didn't perform as well.

**Comment by kirillzyusko** [comment]
> Does that answer your question?

Yes, thank you 🙏 

We've investigated `SharedArrayBuffer` usage, but since it operates on bytes level, we anyway need to serialize/deserialize a data and here we will loose our performance. I. e. if we do `JSON.stringify` to pass a data to worker, then here we are wasting a lot of CPU resources (but if there is another optimal way of converting JS object into `SharedArrayBuffer` - let me know).

We also looked at the performance when we have no serialisation/deserialisation on the worker thread API, but assuming that all the data layer code would operate in the worker thread directly (ie. API calls happening there and `onyx` code lives there as well). For that case we have these results:

|Operation       |SQLite (worker) - sql queries execution time|
|----------------|--------|
|Single set      |21.9|
|Get collection  |9.21|
|Merge collection|229.6|
|Single merge    |124.1|
|Clear           |506.2|

I. e. if there is no communication between threads, then `SQLiteWASM` is an absolute winner 💪 (I also added these numbers to main comparison table).

Let me know what do you about the results and further direction 😊

**Comment by sgbeal** [comment]
> We've investigated SharedArrayBuffer usage, but since it operates on bytes level, we anyway need to serialize/deserialize a data and here we will loose our performance. I. e. if we do JSON.stringify to pass a data to worker, then here we are wasting a lot of CPU resources (but if there is another optimal way of converting JS object into SharedArrayBuffer - let me know).

AFAIK there is no more optimal approach for general-purpose or high-level software. We internally use a fit-to-purpose serialization approach which has proven to be faster, but it's in no way flexible. It depends, for example, on having a fixed number of fixed-size values, which isn't practical for anything but the most specialized of use cases.

> Let me know what do you about the results and further direction

Unfortunately, my lack of background with your apps, infrastructure, and use cases leaves me unable to say terribly much which is likely to be of use.

If you like, we can do a video call and, with some hand-holding from you, perhaps i'll be able to offer some concrete suggestions and/or provide guidance on getting the most out of sqlite's JS/WASM APIs. Just send a meeting link and time to support at sqlite org and i can make the time slot work.

**Comment by sgbeal** [comment]
> I sent an invitation to your email (took it from github profile,

Got it. See you on Friday!

**Comment by sgbeal** [comment]
> VFS-2 can be used with concurrency, but it's supported only in latest Chrome browser versions 

To clarify: that's capability is currently hypothetical. The new Chrome feature _might_ allow us to add concurrency to that VFS, but we won't know for sure until we're able to experiment with it. Even if we do that, however, that OPFS feature is experimental and currently only in Chrome, so it may still change in incompatible ways. We have no information about whether the other browsers plan to adopt that feature and we are not keen on the idea of releasing a Chrome-only VFS.

**Comment by roryabraham** [comment]
most recent discussion [here](https://expensify.slack.com/archives/C05LX9D6E07/p1708591196763269)

**Comment by roryabraham** [comment]
conversation is ongoing, with some recent highlights being:

@quinthar:

> So there are kind of three different optimizations being discussed, and they are entirely independent (ie, we can do any of them by themselves), but we are talking about them as fi they are somehow the same:
>
> 1. Query speed: Use on-disk imaging to improve disk read times
> 2. Worker offload: Add APIs to postprocess (ie, sort/filter) results on the worker thread to improve main thread responsiveness
> 3. Worker overhead: Reduce the amount of data going over postMessage to improve overall command performance
>
> Of all those, I think the most important goal should be offload both querying and postprocessing to the worker thread to keep the main thread continuously responsive.  Improving the query speed is nice, but kind of marginal.

@adhorodyski:

> On the 3 extracted topics, worker offload, worker overhead, query speed (thanks for this clarification!):
I gave it another spin and and as crazy as it gets, I think I've managed to solve all of them at once with the current key-value data structure for SQLite (amazes my every day how powerful it is :stuck_out_tongue:). You can think of it as an API for heavy processing on the other thread with very little adjustments to what we have right now :slightly_smiling_face:
I'll do my best to share a POC shortly with you but need to run some more tests.
Without moving anything around I'm able to query the KV structure using SQL and get the UI reactivity based on the database updates. So far it looks like it can be both performant and require little to changes, just adding the new queries on top and syncing with database hooks

excited to hear more about @adhorodyski's proposal.

**Comment by kidroca** [comment]
I’ve been deeply invested in ongoing discussions about performance and have observed that our primary challenge doesn't seem to stem from the storage database itself but rather from extensive processing times on the main thread. This challenge manifests in two main areas:
- Substantial computations within client code, notably in components like LHN.
- Onyx’s architecture, which processes everything on the main thread, contributing to bottlenecks.

It’s clear these issues are significant and not directly related to the choice of storage provider. I’m curious about the native side - do we face similar challenges there, or is the conversation about moving storage handling off the main thread mainly focused on web implementations?

Given my history with Onyx, I have immense respect for what we've built and its evolution from a simple pub-sub model to a sophisticated key/value storage solution. As we navigate these performance challenges, I believe it's beneficial to revisit the foundational goals of Onyx:
- What needs did Onyx initially aim to fulfill?
- Could exploring a facade or middleware approach with another state management solution, like Redux, benefit us?

### Reflections on Onyx’s Purpose:
- Onyx’s simplicity and the ease of its pub-sub interface for components have been its strengths, enabling quick onboarding and straightforward usage.
- It was designed to function outside of React, which, while initially seen as an advantage, has led to challenges with static connections and the inability to lazy load resources as one might with React components.
- It aimed to reduce boilerplate code.

### Considering Alternatives:
- Before Onyx, had we explored using existing solutions like Redux? Redux’s adaptability with various storage backends could offer valuable insights into alternative architectures.
- The interface Onyx provides for component subscriptions isn’t vastly different from Redux’s. However, the approach to state updates and management diverges significantly, without the structured flow of reducers and actions seen in Redux.

I once hesitated to use Redux, favoring more familiar patterns like MVC with MobX. Yet, the potential for integrating existing solutions through middleware, possibly enhancing Onyx or even adopting new approaches, is worth considering. This doesn’t diminish Onyx’s value but opens a dialogue on leveraging collective wisdom and advancements from the wider community.

As we ponder the future of Onyx, I see an opportunity for evolution rather than abandonment. Could we envision a redesign or integration that honors Onyx's original principles while addressing our current challenges? I’m eager to contribute to a path forward that builds on our collective achievements and addresses the needs of our evolving technology landscape.

**Comment by roryabraham** [comment]
An interesting library we could consider (or pull from for inspiration) is [legend-state](https://legendapp.com/open-source/state/intro/introduction/). Reviewing the design goals for Onyx written by @tgolen, and adding this one:

- it should be as performant as possible

legend-state seems to check all the boxes. I believe it's quite similar to Onyx in its design, with even less boilerplate. I also think it would be feasible to leverage it to simplify Onyx internals while keeping the existing interface for Onyx largely unchanged, avoiding the need for a major app-wide refactoring effort.

I've also met the maintainer in person a few times at React Native conferences, and I believe he'd be open to working with us. A direct line to support could alleviate concerns with lack of control over the library.

**Comment by adhorodyski** [comment]
@roryabraham I'm going to need some more time to prepare a draft of how I think this could look like, but to give you all more context I genuinely think nearly all we need is the right **design** and **composition**, not much of the new tools - this might be the 10% but mostly due to the ergonomics of not rewriting community libraries.

I basically work with **SQL queries & JS Promises**, binding them to the UI.

I'll do my best to share a WIP proposal of this design next week.

**Comment by roryabraham** [comment]
It seems like conversation has stalled for the last few weeks, and the initial investigations we did have prettymuch concluded. It sounds like @adhorodyski has an idea he'd like to propose for which SQLite on the web would be a prerequisite, so I'm going to assign this issue to him. I'm also going to unassign @hannojg and @kirillzyusko because their explorations are complete.

**Comment by adhorodyski** [comment]
@roryabraham still working on it - correct with the web support being a prerequisite, but after a few discussions with @jbroma together we decided to slightly pivot as IndexedDB is for now the only option which fully supports this platform. I'm aiming for a design where once web catches up, we can easily migrate just this very piece with a very little effort. Will update within days!

**Comment by roryabraham** [comment]
Hey @adhorodyski, we [discussed this internally](https://expensify.slack.com/archives/C03TQ48KC/p1711477685382299), and since we're not working from a clear problem statement here and the ROI of any changes is still very much up-in-the-air, we've agreed that we're going to close this issue for now.

We advise that you shift your focus into other areas of the performance audit for the time being. We love SQLite and may want to pick this discussion back up someday, but for now we're going to cap it here for #focus. Thanks everyone for all your participation so far!

**Comment by adhorodyski** [comment]
@roryabraham I agree this was fine to be closed - I'll post my research on Slack just so we don't loose it later on, I think this is still valuable and holds/maybe we'll continue the discussion from there/kill it with fire for now :D

**Comment by adhorodyski** [comment]
cc @roryabraham posted internally under [this thread](https://callstack-hq.slack.com/archives/C05LX9D6E07/p1712068782624469), please let me know if there's a better place for this :)

**Comment by roryabraham** [comment]
thanks, I think that's an appropriate place for that
